### ISSUE TITLE
Replace does not act correctly with part of the current version

### ISSUE THREAD

**Issue by Sytten** [c4urself/bump2version#71]
Given a `setup.py` file containing those two lines:
```python
if pkg_resources.get_distribution("setuptools").version < "40.1.1":

# ... other lines

    version="0.1.1",
```

If you bump the version (the patch for example), it will modify `40.1.1` to `40.1.2`. This is because it's a simple replace using the current version. It probably should use a regex to find the version the file. The only work around I found was to the change the search to include the double quotes.

**Comment by florisla** [comment]
How does adding double quotes to `search` help here?  `40.1.1` and `0.1.1` both have double quotes around them...

Was your intent to bump `40.1.1` to `40.1.2` or to bump `0.1.1` to `0.1.2` ?

Can you show your bumpversion configuration file and the command you use?

**Comment by Sytten** [comment]
This has been a while and I don't have access to the code anymore (proprietary).
The goal was to bump `0.1.1` to `0.1.2` without changing the `40.1.1`.
In this example the double quotes help because if you only search for `0.1.1` it will find 2 elements `4[0.1.1]` and `[0.1.1]`. If you search for `"0.1.1"`, it will only match the version number (that we want to modify).

**Comment by florisla** [comment]
I see, you got an coincidental match.

The `--current-version` argument can't protect against that.

For future reference, you can use a custom `search` that only matches what you want:

    [bumpversion]
    current_version = 0.1.1

    [bumpversion:file:setup.py]
    search = "{current_version}"
    replace = "{new_version}"

**Comment by Sytten** [comment]
I know, but I don't agree with your solution. If the regex for search was correct it wouldn't match 40 with 0 and thus not replace the version (for that I believe it needs to have a greedy search).

**Comment by florisla** [comment]
Got it, the `search` matches `40.1.1` and 0.1.1; only the latter is recognized as the current version number.   But the naive replacement code does two replacements instead of one.

**Comment by Sytten** [comment]
Yes exactly! It's probably not a big change, but we need to ensure backward compatibility with tests.

**Comment by florisla** [comment]
This is not very trivial to fix since the current `utils.ConfiguredFile.replace()` does not use regular expressions (`re.sub`).  It's simply serializing the `search` and `replace` text, and doing `str.replace` on the whole file content.

**Comment by florisla** [comment]
I'm not working on a fix, but here is a unit test which demonstrates the problem.

https://github.com/florisla/bump2version/tree/dont-replace-partial-matches

**Comment by florisla** [comment]
I'm trying to use `re.sub` and adding leading and trailing `\b` (word boundary) in the expression.

This works fine preventing the issue, but it has nasty side effect.

Previously, it would happily replace 'v0.1.1' with 'v0.1.2'.  The word boundary checking prevents that, and requires a custom `search` and `replace` with the v prefix defined.

**Comment by panicgh** [comment]
Maybe I miss some context, but what about matching `[^\d]` ("not a decimal digit", same as `\D`), instead of `\b` (word boundary)? In addition, accept start of line and end of line, in case there is no prefix/suffix.

For example:

```python
def ver_replace(current_version, new_version, string):
    repl = r'\g<prefix>' + new_version + r'\g<suffix>'
    current_version = re.escape(current_version)
    pattern = r'(?P<prefix>[^\d]|^)' + current_version + r'(?P<suffix>[^\d]|$)'
    return re.sub(pattern, repl, string)
```

```python
>>> ver_replace('1.2.3', '1.2.4', r'1.2.3')
'1.2.4'          # correctly replaced
>>> ver_replace('1.2.3', '1.2.4', r'v1.2.3')
'v1.2.4'         # correctly replaced with prefix
>>> ver_replace('1.2.3', '1.2.4', r'vvvv1.2.3')
'vvvv1.2.4'   # correctly replaced with multi-char prefix
>>> ver_replace('1.2.3', '1.2.4', r'91.2.3')
'91.2.3'         # correctly NOT replaced as it is wrong version
>>> ver_replace('1.2.3', '1.2.4', r'1.2.33')
'1.2.33'         # correctly NOT replaced as it is wrong version
>>> ver_replace('1.2.3', '1.2.4', r'1.2.3v')
'1.2.4v'         # correctly replaced even with suffix
>>> ver_replace('1.2.3', '1.2.4', r'abc\n1.2.3\ndef')
'abc\\n1.2.4\\ndef'    # correctly replaced even in case of newline
```
